client/attest.go

```diff
@@ -39,74 +38,74 @@ type AttestOpts struct {
 	FetchCertChain bool
 }
 
-// Given a certificate, iterates through its IssuingCertificateURLs and returns the certificate
-// that signed it. If unable to find an intermediate certificate, it returns a nil.
-func fetchIssuingCertificate(cert *x509.Certificate) *x509.Certificate {
-	if cert == nil {
-		return nil
+// Given a certificate, iterates through its IssuingCertificateURLs and returns
+// the certificate that signed it. If the certificate lacks an
+// IssuingCertificateURL, return nil. If fetching the certificates fails or the
+// cert chain is malformed, return an error.
+func fetchIssuingCertificate(cert *x509.Certificate) (*x509.Certificate, error) {
+	// Nothing to do
+	if cert == nil || len(cert.IssuingCertificateURL) == 0 {
+		return nil, nil
 	}
 	// For each URL, fetch and parse the certificate, then verify whether it signed cert.
 	// If successful, return the parsed certificate. If any step in this process fails, try the next url.
+	// If all the URLs fail, return an error.
+	var certErr error
 	for i, url := range cert.IssuingCertificateURL {
 		// Limit the number of attempts.
-		if i == maxIssuingCertificateURLs {
-			log.Printf("Reached the maximum number of attempts (%v)", maxIssuingCertificateURLs)
-			return nil
+		if i >= maxIssuingCertificateURLs {
+			break
 		}
 		resp, err := http.Get(url)
 		if err != nil {
-			log.Printf("failed to retrieve certificate at %v: %v\n", url, err)
+			certErr = fmt.Errorf("failed to retrieve certificate at %v: %w", url, err)
```

Comment:
This (and other err assignments below) could be overwritten by later loops. In the event of reaching max attempts only the last certErr would be returned. Is this the intended behavior? Why not append the errors so it doesn't get overwritten (ex. in the event the important failure point isn't the last certificate attempted)

Reply:
The overwriting here is intentional (and I've updated the code to reflect this). Eventually, we should return multiple errors, but it's probably easier to deal with that when we fix #169.

At the end of the day, it probably doesn't really matter. The vast, vast majority of certs will either have 0 or 1 URLs in `Cert.IssuingCertificateURL`, so it doesn't really matter.